### ISSUE TITLE
add keyword which refers to thing in most immediate scope

### ISSUE THREAD

**Issue by andrewrk** [andrewrk/zig#169]
Proposed keyword: `this`

It refers to whatever thing is in most immediate scope.

Example:

```zig
const module = this;

fn factorial(x: i32) {
    if (x == 1) x else x * this(x)
}

const Node = struct {
    item: i32,
    next: &this,
};

struct MersenneTwister(
    int: type, n: usize, m: usize, r: int,
    a: int,
    u: int, d: int,
    s: int, b: int,
    t: int, c: int,
    l: int, f: int)
{
    const Self = this;

    array: [n]int,
    index: usize,

    pub fn init(self: &Self, seed: int) {
        module.factorial(seed);
        // ...
    }
}
```

**Comment by phase** [comment]
> Only having one way to do things is one of Zig's goals.

```zig
fn factorial(x: i32) {
    if (x == 1) x else x * this(x)
}
```

can already be done like

```zig
fn factorial(x: i32) {
    if (x == 1) x else x * factorial(x)
}
```

so there's no need to make another option. Structs could use the name of the type instead of declaring a variable.

```zig
struct MersenneTwister {
    // ...
    pub fn init(self: &MersenneTwister, seed: int) {
        module.factorial(seed);
        // ...
    }
}
```

**Comment by andrewrk** [comment]
The important example here is the anonymous struct:

```zig
const Node = struct {
    item: i32,
    next: &this,
};
```

We need this feature for #160 to work.

The other examples above I don't think would be idiomatic, just examples of the semantics of this feature.

Another example is https://github.com/andrewrk/zig/blob/0ae9023832584e256aa3b6df0f0829026141d21a/std/cstr.zig#L7

This `const strlen = len;`. Inside CBuf, `len` refers to a method, so you can't call the global `len`. If we did earlier `const module = this;` then you could do `module.len`.

**Comment by andrewrk** [comment]
Another use case for this is to make the language smaller and make annotations more powerful. For example, if we allow a block at the top level scope, we could do function annotations like this:

```
{@markAsTest(functionName)}
fn functionName() {
    // do the test...
}
```

With `this` you could potentially do this annotation inside the function:

```
fn functionName() {
    @markAsTest(this);
    // do the test
}
```

More importantly though we could have block-level attributes like this:

```
fn someFunction() {
    // some normal code...
    {
        @setDebugSafety(this, false);
        // some fast unsafe code
    }
    // some more normal code...
}
```

Here, `this` refers to a block, and `setDebugSafety` builtin function disables debug safety mechanisms only for the block that it is in.

With this strategy we could get rid of the compiler directive syntax and make the language smaller.

**Comment by thejoshwolfe** [comment]
what would `@typeOf({this})` be? a block scope metatatype instantiation? what would be the members of `{this}`? perhaps `.parent`? perhaps `.declarations` which contains a member for each var/const/ect declared in that scope? if all this sounds good, then the factorial example above is inconsistent. `this` in a function should not refer to the function, but the main block scope in the function body. this means you can't `this(x - 1)`, so that resolves the rule of orthogonality violation.

but also, what the hell is the point of the features i just described? do we really need all this? if all that is the cost of removing annotations, i think the goal of making the language smaller fails. what's really the problem we're trying to solve with this issue?

for the `MersenneTwister` example (the original one) the benefit is not having to repeat all the template parameters if you want to refer to your own type. this is a good thing imo, because I'm still shuddering from the horrors of function signature duplication between .c and .h files. but a competing proposal for this usecase is to simply define `Self` in every struct declaration implicitly. (i don't believe that proposal has an issue number yet.)

another usecase is the `const strlen = len;` example. I've been arguing all along that the problem there is with the names, not with the language. i think the so-called `len` function should either be called `strlen`, a reference to c, or how about `findNullTerminator`. the second example follows the pattern (which i don't know if I've seen explicitly spelled out) that functions should be verbs unless they're O(1) getters, in which case they can be nouns.

in short, what (possibly) compelling usecases do we really have for `this` issue?

**Comment by andrewrk** [comment]
> but also, what the hell is the point of the features i just described? do we really need all this?

I don't think we want all the features you described. I think `@typeOf({this})` would be `(block)` and doing pretty much anything with that type other than calling a few builtin functions on it would be a compile error.



> what's really the problem we're trying to solve with this issue?
> in short, what (possibly) compelling usecases do we really have for `this` issue?

 * Get reference to container type inside it without it having to explicitly name it so that anonymous types can work. In this regard `this` is like the `Self` proposal but more general purpose, so that the additional following use cases are solved.
 * Annotations can work with existing builtin function syntax instead of `#attribute(...)` special syntax.
 * Annotations can target blocks of code instead of only functions and declarations, and participate in logic constructs such as an if-then statement with a condition on a compile time variable.
 * Ability to refer to module-level declarations when shadowed by local declarations. (strlen is a bad example since the validity of the name is in question, but I claim that the issue still stands)

 Your point seems to be that `this` drags in all this other meta programming baggage, but I don't see why that has to be the case.

**Comment by thejoshwolfe** [comment]
ok, good points. I'm probably doing a little too much exploration, but you brought up another point that i missed: safety and other annotations having conditional logic. here goes some more exploration!

is it possible to refer to a block while outside it? i believe the syntax as proposed makes that only possible if the block is the return value from the block (`const block = {foo(); this};`), which means you'd probably want to always annotate your blocks inside the block, instead of after the block, and therefore we might want to make this a restriction to have more orthogonality. this restriction would come in the form of forbidding a block value to escape it's block scope.

naming a block value into a const is still useful if you're conditional logic wants to be written with more blocks; for example:

```
fn foo() {
  var x: u32 = 100;
  while (x != 0) {
    const block = this;
    if (gotta_go_fast) {
      @annotate(block, "unsafe");
      if (is_posix) {
        disableInterrupts();
      }
    }
    print(x);
    x -= 1;
  }
}
```

**Comment by andrewrk** [comment]
More exploration!

```zig
fn foo() {
    const block_result = {
        const outer_block = this;
        if (blah) {
            while (true) {
                const outer_while = this;
                while (true) {
                    if (something) {
                        break outer_while;
                    } else {
                        @returnFromBlock(outer_block, 1234);
                    }
                }
            }
        } else {
            5678
        }
    };
}
```
